### PR TITLE
Finetune

### PR DESCRIPTION
**Pull request by EthanReid**
Example scripts to finetune the text and region model.

### CODE REVIEW

File: moondream/finetune/finetune_region.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+    s_loss = F.cross_entropy(s_logits, s_labels)
+
+    return c_loss + s_loss
+
+
+class CocoDataset(Dataset):
+    """
+    Dataset class for COCO type data.
+    To download the Roboflow railwayvision dataset visit: https://universe.roboflow.com/research-zwl99/railwayvision
```
Comment (parsakhaz) [inline moondream/finetune/finetune_region.py:73]: Coco JSON format

File: moondream/finetune/finetune_region.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    optimizer = AdamW8bit(
+        [
+            {"params": model.region.parameters()},
+        ],
+        lr=LR,
+        betas=(0.9, 0.95),
+        eps=1e-6,
+    )
+
+    # Add path to annotation file and img dir
```
Comment (parsakhaz) [inline moondream/finetune/finetune_region.py:169]: Remove hardcoded paths

**Comment by forensicmike** [comment]
This is fantastic @EthanReid thank you!!

One question, is it a problem that we are not calling `model.text.train()` somewhere? 

2nd, should we possibly be using the `config.tokenizer.templates` instead of hard-coding? My thought process is that if we want to train on captioning, for example, fine-tuning with the `\n\nQuestion:Describe this image.\n\nAnswer:` doesn't seem to have much impact on the caption answers.

File: moondream/finetune/finetune_text.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+from bitsandbytes.optim import AdamW8bit
+import wandb
+
+from ..torch.weights import load_weights_into_model
+from ..torch.moondream import MoondreamModel, MoondreamConfig, text_encoder
+from ..torch.text import _produce_hidden, _lm_head, TextConfig
+
+MODEL_PATH = ""
+ANSWER_EOS = "<|endoftext|>"
+LR = 1e-5
```
Comment (parsakhaz) [inline moondream/finetune/finetune_text.py]: Update to:
LR = 5e-6
EPOCHS = 3
GRAD_ACCUM_STEPS = 128

File: moondream/finetune/finetune_text.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    )
+
+    dataset = DocciDataset("train")
+
+    total_steps = EPOCHS * len(dataset) // GRAD_ACCUM_STEPS
+    pbar = tqdm(total=total_steps)
+
+    i = 0
+    for epoch in range(EPOCHS):
+        for sample in dataset:
```
Comment (forensicmike) [inline moondream/finetune/finetune_text.py:107]: Should we be shuffling the dataset between epochs? The dataloader from the notebook did this automatically.

Comment (EthanReid) [reply moondream/finetune/finetune_text.py:107]: Yep, thanks for the feedback; that would be a good thing to add into your script. Regarding this script, it is intended to serve as a barebones starting point.
